toolchain/check/scope_stack.h

```diff
@@ -0,0 +1,195 @@
+
+    // The value corresponding to the current `returned var`, if any. Will be
+    // set and unset as `returned var`s are declared and go out of scope.
+    SemIR::InstId returned_var = SemIR::InstId::Invalid;
+  };
+
+  // A non-lexical scope in which unqualified lookup may be required.
+  struct NonLexicalScope {
+    // The index of the scope in the scope stack.
+    ScopeIndex scope_index;
+
+    // The corresponding name scope.
+    SemIR::NameScopeId name_scope_id;
+  };
+
+  // Pushes a scope onto scope_stack_. NameScopeId::Invalid is used for new
+  // scopes. lexical_lookup_has_load_error is used to limit diagnostics when a
+  // given namespace may contain a mix of both successful and failed name
+  // imports.
+  auto Push(SemIR::InstId scope_inst_id = SemIR::InstId::Invalid,
+            SemIR::NameScopeId scope_id = SemIR::NameScopeId::Invalid,
+            bool lexical_lookup_has_load_error = false) -> void;
+
+  // Pops the top scope from scope_stack_, cleaning up names from
+  // lexical_lookup_.
+  auto Pop() -> void;
+
+  // Pops the top scope from scope_stack_ if it contains no names.
+  auto PopIfEmpty() -> void {
+    if (scope_stack_.back().names.empty()) {
+      Pop();
+    }
+  }
+
+  // Pops scopes until we return to the specified scope index.
+  auto PopToScope(ScopeIndex index) -> void;
+
+  // Looks up the name `name_id` in the current scope. Returns the existing
+  // lookup result, if any.
+  auto LookupInCurrentScope(SemIR::NameId name_id) -> SemIR::InstId;
+
+  // Looks up the name `name_id` in the current scope and its enclosing scopes.
+  // Returns the innermost lexical lookup result, if any, along with a list of
+  // non-lexical scopes in which lookup should also be performed, ordered from
+  // outermost to innermost.
+  auto LookupInEnclosingScopes(SemIR::NameId name_id)
+      -> std::pair<SemIR::InstId, llvm::ArrayRef<NonLexicalScope>>;
+
+  // Looks up the name `name_id` in the current scope. Returns the existing
+  // instruction if any, and otherwise adds the name with the value `target_id`
+  // and returns Invalid.
+  auto LookupOrAddName(SemIR::NameId name_id, SemIR::InstId target_id)
+      -> SemIR::InstId;
+
+  // Returns the scope index associated with the current scope.
+  auto current_scope_index() const -> ScopeIndex {
+    return current_scope().index;
+  }
+
+  // Returns the name scope associated with the current lexical scope, if any.
+  auto current_scope_id() const -> SemIR::NameScopeId {
+    return current_scope().scope_id;
+  }
+
+  // Returns the instruction associated with the current scope, or Invalid if
+  // there is no such instruction, such as for a block scope.
+  auto current_scope_inst_id() const -> SemIR::InstId {
+    return current_scope().scope_inst_id;
+  }
+
+  // Returns the current scope, if it is of the specified kind. Otherwise,
+  // returns nullopt.
+  template <typename InstT>
+  auto GetCurrentScopeAs(const SemIR::File& sem_ir) -> std::optional<InstT> {
```

Comment:
Similar to `PushScope` -> `Push`, I think you could probably rename `current_scope` -> `current` or `peek`

Reply:
Done.